source/slang/slang.cpp

```diff
@@ -1318,6 +1321,63 @@ SLANG_NO_THROW SlangResult SLANG_MCALL Linkage::createCompileRequest(
     return SLANG_OK;
 }
 
+SLANG_NO_THROW SlangResult SLANG_MCALL Linkage::computeDependencyBasedHash(
+    SlangInt targetIndex,
+    slang::Checksum* outHash)
+{
+    slang::Checksum hash;
+    MD5Context context;
+    MD5HashGen hashGen;
+    hashGen.init(&context);
+
+    // Add the Slang compiler version to the hash
+    auto version = String(getBuildTagString());
+    hashGen.update(&context, version);
+
+    // Add the search directory paths to the hash
+    auto searchDirectoryList = getSearchDirectories().searchDirectories;
+    for (auto& searchDir : searchDirectoryList)
+    {
+        auto searchPath = searchDir.path;
+        hashGen.update(&context, searchPath);
+    }
+
+    // Add the preprocessor definitions to the hash
+    for (auto& key : preprocessorDefinitions)
+    {
+        hashGen.update(&context, key.Key);
+        hashGen.update(&context, key.Value);
+    }
+
+    // Add the target specified by targetIndex
+    auto targetReq = targets[targetIndex];
+    hashGen.update(&context, targetReq->getTarget());
+    hashGen.update(&context, targetReq->getTargetFlags());
+    hashGen.update(&context, targetReq->getFloatingPointMode());
+    hashGen.update(&context, targetReq->getLineDirectiveMode());
+    hashGen.update(&context, targetReq->shouldDumpIntermediates());
+    hashGen.update(&context, targetReq->getForceGLSLScalarBufferLayout());
+    hashGen.update(&context, targetReq->shouldTrackLiveness());
+
+    auto targetProfile = targetReq->getTargetProfile();
+    hashGen.update(&context, targetProfile.getStage());
+    hashGen.update(&context, targetProfile.getVersion());
+    hashGen.update(&context, targetProfile.getFamily());
+
+    auto targetProfileName = String(targetProfile.getName());
+    hashGen.update(&context, targetProfileName);
+
+    auto cookedCapabilities = targetReq->getTargetCaps().getExpandedAtoms();
+    for (auto& capability : cookedCapabilities)
+    {
+        hashGen.update(&context, capability);
```

Comment:
Thinking out loud: there is a fair bit of subtlety here, where some of the fields we are including in the hash are stored as an `enum` that is *not* guaranteed to use the exact same encoding across different builds of the compiler. That situation sounds worrying ("what if a new version of the compiler uses the same integer value to represent a different capability?"), but it is actually okay in practice for the very specific reason that we are also including the compiler version into the hash, so different compiler versions would already be hashing different data at an earlier point, so that overlapping capability encodings (for example) would not create a problem.

Reply:
I've added a comment about this in the new `ComponentType::computeDependencyBasedHash` implementation